pudb/var_view.py

```diff
@@ -386,84 +470,25 @@ def walk_value(self, parent, label, value, id_path=None, attr_prefix=None):
-                                cont_id_path)
-                            break
-
-                    self.walk_value(new_parent_item, None, entry,
-                        "%s[%d]" % (id_path, i))
-                if not value:
-                    self.add_item(new_parent_item, "<empty>", None)
-                return
-
             # containers --------------------------------------------------
-            key_it = None
-
-            try:
-                key_it = value.keys()
-            except AttributeError:
-                # keys or iterkeys doesn't exist, not worth mentioning!
-                pass
-            except Exception:
-                ui_log.exception("Failed to obtain key iterator")
-
-            if key_it is None:
-                try:
-                    len_value = len(value)
-                except (AttributeError, TypeError):
-                    # no __len__ defined on the value, not worth mentioning!
-                    pass
-                except Exception:
-                    ui_log.exception("Failed to determine container length")
-                else:
-                    try:
-                        value[0]
-                    except (LookupError, TypeError):
-                        key_it = []
-                    except Exception:
-                        ui_log.exception("Item is not iterable")
-                    else:
-                        key_it = xrange(len_value)
-
-            if key_it is not None:
-                cnt = 0
-                for key in key_it:
-                    if cnt % 10 == 0 and cnt:
-                        cont_id_path = "%s.cont-%d" % (id_path, cnt)
-                        if not self.frame_var_info.get_inspect_info(
-                                cont_id_path, read_only=True).show_detail:
-                            self.add_item(
-                                new_parent_item, "...", None, cont_id_path)
-                            break
-
-                    try:
-                        next_value = value[key]
-                    except (LookupError, TypeError):
-                        ui_log.exception("Failed to iterate an item that "
-                            "appeared to be iterable.")
-                        break
-
-                    self.walk_value(new_parent_item, repr(key), next_value,
-                        "%s[%r]" % (id_path, key))
-                    cnt += 1
-                if not cnt:
-                    self.add_item(new_parent_item, "<empty>", None)
+            if isinstance(value, PudbMapping):
+                self.walk_mapping(new_parent_item, label, value, id_path)
+                return
+            elif isinstance(value, PudbSequence):
+                self.walk_sequence(new_parent_item, label, value, id_path)
+                return
+            elif isinstance(value, PudbCollection):
+                self.walk_collection(new_parent_item, label, value, id_path)
```

Comment:
Would `singledispatch` work for this? (AKA how dependent are we on the order of these tests?)

Reply:
With the way the ABCs are set up right now there's an ordering component: a mapping is a sequence which also has `keys` defined, and either one is likely to be recognized as a collection.